src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs

```diff
@@ -834,43 +836,81 @@ internal unsafe int GetNonRandomizedHashCode()
             }
         }
 
-        // Use this if and only if 'Denial of Service' attacks are not a concern (i.e. never used for free-form user input),
-        // or are otherwise mitigated
-        internal unsafe int GetNonRandomizedHashCodeOrdinalIgnoreCase()
+        internal int GetNonRandomizedHashCodeOrdinalIgnoreCase()
         {
-            fixed (char* src = &_firstChar)
-            {
-                Debug.Assert(src[this.Length] == '\0', "src[this.Length] == '\\0'");
-                Debug.Assert(((int)src) % 4 == 0, "Managed string should start at 4 bytes boundary");
+            return GetNonRandomizedHashCodeOrdinalIgnoreCaseStatic(ref _firstChar, Length, true);
+        }
 
-                uint hash1 = (5381 << 16) + 5381;
-                uint hash2 = hash1;
+        private static int GetNonRandomizedHashCodeOrdinalIgnoreCaseStatic(ref char firstChar, int length, bool normalizeNonAscii)
+        {
+            Debug.Assert(Unsafe.Add(ref firstChar, length) == '\0', "*(&firstChar + length) == '\\0'");
 
-                uint* ptr = (uint*)src;
-                int length = this.Length;
+            uint hash1 = (5381 << 16) + 5381;
+            uint hash2 = hash1;
 
-                // We "normalize to lowercase" every char by ORing with 0x0020. This casts
-                // a very wide net because it will change, e.g., '^' to '~'. But that should
-                // be ok because we expect this to be very rare in practice.
+            // We "normalize to lowercase" every char by ORing with 0x0020. This casts
+            // a very wide net because it will change, e.g., '^' to '~'. But that should
+            // be ok because we expect this to be very rare in practice.
 
-                const uint NormalizeToLowercase = 0x0020_0020u; // valid both for big-endian and for little-endian
+            const uint NormalizeToLowercase = 0x0020_0020u; // valid both for big-endian and for little-endian
 
-                while (length > 2)
+            int i = 0;
+            int count = length;
+
+            while (count > 2)
+            {
+                ref byte firstByte = ref Unsafe.As<char, byte>(ref firstChar);
+                uint p0 = Unsafe.ReadUnaligned<uint>(ref Unsafe.Add(ref firstByte, i));
+                uint p1 = Unsafe.ReadUnaligned<uint>(ref Unsafe.Add(ref firstByte, i + 4));
+                if (normalizeNonAscii && !Utf16Utility.AllCharsInUInt32AreAscii(p0 | p1))
```

Comment:
Nit: if perf tests show a regression, we can hoist the _normalizeNonAscii_ check out of the loop, duplicating the loop. But probably not worth it unless we get evidence of regression.

```cs
if (normalizeNonAscii) { while (count > 2) { /* ... */ } }
else { while (count > 2) { /* ... */ } }
```

Reply:
Sure, will check. Btw, this type of optimization might eventually appear in JIT I guess since it's mentioned in https://github.com/dotnet/runtime/issues/43549 (Loop Unswitching)